### PR TITLE
Semesters singleton migration

### PR DESCRIPTION
**Pull request by andxu282**
### Summary 

Begins the migration script to move semesters to a list of semesters by writing to a new field called plans that has a list of a list of semesters for each user.
- [x] Added field called plans and makes it so that it is written to whenever semesters is written to 
- [x] Read from plans whenever reading from semesters

Remaining TODOs:

- [ ] Run migration script in dev/prod
- [ ] Remove semester field
- [ ] Run remove script in dev/prod

### CODE REVIEW

File: src/firebase-config-common.ts [hunk 1]
```diff
@@ -16,6 +16,7 @@ export const getTypedFirestoreDataConverter = <T>(): CommonFirestoreDataConverte
 });
 
 export type SemesterDocumentData = {
```
Comment (zachary0kent) [inline src/firebase-config-common.ts:18]: Can we define this as `Readonly<SemesterDocumentData>` where `SemesterDocumentData` refers to the mutable type defined in `src/firebase-admin-config`?

File: src/requirements/admin/semesters-singleton-list-migration.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+import { usernameCollection, semestersCollection } from '../../firebase-admin-config';
+
+/**
+ * Perform migration of semester to plans with a list of semesters
+ */
+async function runOnUser(userEmail: string) {
+  const semestersDoc = await semestersCollection.doc(userEmail).get();
+
+  const semesters = semestersDoc.data()?.semesters ?? [];
+
```
Comment (zachary0kent) [inline src/requirements/admin/semesters-singleton-list-migration.ts:12]:
```suggestion
  const semesters = semestersDoc.data()?.semesters ?? [];
```

Comment (zachary0kent) [reply src/requirements/admin/semesters-singleton-list-migration.ts:12]: Spacing looks a bit awkward

File: src/firebase-config-common.ts (hunk 1 above)
Comment (benjamin-shen) [reply src/firebase-config-common.ts:18]: I changed this in #724 btw

File: src/requirements/admin/semesters-singleton-list-migration.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  }
+  const collection = await usernameCollection.get();
+  const userEmails = collection.docs.map(it => it.id);
+  for (userEmail of userEmails) {
+    console.group(`Running on ${userEmail}...`);
+    // Intentionally await in a loop to have no interleaved console logs.
+    // eslint-disable-next-line no-await-in-loop
+    await runOnUser(userEmail);
+    console.groupEnd();
+  }
```
Comment (zachary0kent) [inline src/requirements/admin/semesters-singleton-list-migration.ts:30]:
```suggestion
  const userEmail = process.argv[2];
  if (userEmail != null) {
    await runOnUser(userEmail);
    return;
  }
  const collection = await usernameCollection.get();
  for (const { id } of collection.docs) {
    console.group(`Running on ${id}...`);
    // Intentionally await in a loop to have no interleaved console logs.
    // eslint-disable-next-line no-await-in-loop
    await runOnUser(id);
    console.groupEnd();
  }
```

**Review by benjamin-shen — COMMENTED** [review]
Nice job, this looks pretty good! In this PR can you also start reading from plans, if the field exists?

Could you also
- flesh out the TODO steps
- include some sample commands in the PR description

File: src/utilities.ts [hunk 4]
```diff
@@ -11,6 +12,13 @@ export const SeasonOrdinal = {
   Fall: 3,
 } as const;
 
+export const getFirstPlan = (semestersDoc: SemesterDocumentData): readonly FirestoreSemester[] => {
+  if (semestersDoc.plans) {
```
Comment (zachary0kent) [inline src/utilities.ts:16]:
```suggestion
  if (semestersDoc.plans && semestersDoc.plans.length > 0) {
```

**Comment by noschiff** [comment]
@zachary0kent what's the status of this PR?

**Review by zachary0kent — APPROVED** [review]
Looks good, I ran the migration script on my dev account and everything works as expected.

**Comment by zachary0kent** [comment]
> @zachary0kent what's the status of this PR?

done

**Pull request merged**